Course navigation menu accessibility issue

clrux opened the pull request:
This work relates to [AC-76](https://openedx.atlassian.net/browse/AC-76) continues where Ahsan left off. It combines styling work I completed earlier (included in this branch) with structural work from Ahsan. According to the previous (now closed) PR the following have been done:
* Change the structure of the course navigation Menu HTML
* Add some basic styling
* Add basic JS for html rendering
* Update tests according to new structure

## Browsers

This work was checked in the following browsers:

| Platform | Browser | Version | Passed |
|----------|---------|---------|--------|
| Mac      | Chrome  | 43      | Yes    |
|          | Safari  | 8       | Yes    |
|          | Firefox | 24      | Yes    |
| Windows  | Chrome  | 43      | Yes    |
|          | Firefox | 38      | Yes    |
|          | IE      | 10      | Yes    |
|          | IE      | 11      | Yes    |

## Sandbox

http://clrux-ac-76.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/

## Various tests and reading

During the course of this work I discovered that Firefox doesn't draw focus outlines the same way as other browsers for buttons. IE, Chrome, and Safari will draw a focus outline on an element, regardless of its styling. Firefox however, will use a different styling or none at all if borders or backgrounds are styled. We conducted [several tests](http://jsfiddle.net/clrux/e9xdod0k/) on this to verify.

To prevent the focus outline from appearing on click (but retaining it on keyboard navigation) I opted to use `a role="button"` instead of `button` for the toggles. Buttons, when clicked, inherently receive focus unless it's sent elsewhere, whereas links do not. [Read more about the `button role`](http://www.w3.org/TR/wai-aria/roles#button).

## Reviewers

The following reviewers need to provide thumbs before this work can be merged.

- [x] @cptvitamin (Accessibility)
- [x] @talbs (FED/Sass)
- [x] @AlasdairSwan (JS)

### Notes to the reviewers

__All__
Sorry for so many pings with various PRs. The first one was accidentally closed, then reverted. Then I opened a second but for some reason only the last five commits made it in. Cherry-picking  the earlier commits only caused issues, so I ended up closing that branch and making yet another PR with everything cherry-picked properly (this one). Please review here.

__Mark__
VoiceOver in Chrome still announces the incorrect number of items, despite converting the `li`'s to normal `a`'s within the `nav`. Otherwise, his reads much clearer than before and has correct focus management.

I moved the `aria-expanded` from the button to the actual container that's expanded. It seems to read okay with screen readers.

__Brian__
Styles should look pretty much like they did before with very few exceptions. I ditched the slide transitions because it wasn't allowing focus to be sent (race condition). Let me know your thoughts on overall styles.

__Alasdair__
I tried to emulate the standards from the dropdown menu we briefly worked on a few months ago. I've re-written the course nav code to be JavaScript/jQuery rather than CoffeeScript and removed all references of the the latter. I've also made this use RequireJS for better management later.

One question I did have was how to properly include this file. It will be included everywhere in courseware (that has sections and units) within the LMS. Advice is welcome and thank you in advance.

clrux commented:
jenkins run js

cptvitamin commented on lms/templates/courseware/accordion.html:
```diff
@@ -6,73 +6,44 @@
+    aria_label = chapter['display_name']
+    active_class = ''
+%>
+<button class="button-chapter chapter ${active_class}" id="${chapter['display_id']}-parent" aria-controls="${chapter['display_id']}-child" aria-pressed="false">
```
@clrux I don't think we should be using aria-checked here since that is typically used in situations where buttons are being used as a toggle, light a light switch, you push one button and the other pops up.  Screen readers report this as being "checked" or not.  lets change aria-pressed to to aria-expanded which more accurately describes the state of these controls.

cptvitamin commented on lms/templates/courseware/accordion.html:
```diff
@@ -6,73 +6,44 @@
+<%
+if chapter.get('active'):
+    aria_label = _('{chapter}, current chapter').format(chapter=chapter['display_name'])
+    active_class = 'active"'
```
@clrux there is an unnecessary quote (") here that is causing issues.

clrux replied:
This is actually okay. It looks different in the actual file and would throw JS errors otherwise.

clrux replied:
Is it causing issues in a screen reader? It doesn't throw JS errors for tests.

andy-armstrong replied:
This is Python code, so I would be surprised if the page rendered at all.

cptvitamin commented on lms/templates/courseware/courseware.html:
```diff
@@ -165,9 +165,6 @@
 
 % if disable_accordion is UNDEFINED or not disable_accordion:
     <div class="course-index" role="navigation">
```
@clrux since there is a <nav> element further down in the DOM, we don't need a role defined here.  Especially since it doesn't have a label.  It can be removed all together.

cptvitamin commented on lms/templates/courseware/accordion.html:
```diff
@@ -6,73 +6,44 @@
+    active_class = ''
+%>
+<button class="button-chapter chapter ${active_class}" id="${chapter['display_id']}-parent" aria-controls="${chapter['display_id']}-child" aria-pressed="false">
+    <h3 class="group-heading ${active_class}" aria-label="${aria_label}">
```
@clrux I think use of a Heading here is inappropriate.  Heading elements should be used to indicate the hierarchy of page content, not necessarily nav content.  I think this will result in too many Headings on the page, which reduces the effectiveness of navigating by Heading.  SR users who navigate by Heading expect to find content related to that heading following it, but here, they just find a list of links to subcontent. It makes sense to make these an ordered list

cptvitamin commented on lms/templates/courseware/accordion.html:
```diff
@@ -6,73 +6,44 @@
+        ${chapter['display_name']}
+    </h3>
+</button>
+<div class="chapter-content-container" tabindex="-1" aria-expanded="false">
```
@clrux is there any reason why we can't consolidate this div with the one below?  I worry that this one has a tabindex-1 and the next has a role of region, meaning they may both result in being reported by SR.  I think we can consolidate all the properties. Bring the role, aria-labelledby and aria-hidden from below up here and get rid of the div below.

andy-armstrong commented on lms/static/coffee/src/navigation.js:
```diff
@@ -0,0 +1,70 @@
+var edx = edx || {},
```
It would be good to move this out of the ``coffee`` directory as this isn't a coffee file anymore. I recommend putting it in ``lms/static/js/components` if it a purely LMS feature. If this is a pattern library component, I'd probably put it in ``common/static/common/js/components``.

clrux replied:
@andy-armstrong This is the file I had to replace because the pipeline was looking for it. Once we remove the requirement from the pipeline, we can remove this file.

andy-armstrong replied:
Ah, that's why I'm not seeing any issues on your branch! I'll remove the file and try again...

clrux replied:
I'm not sure how to remove things from the pipeline, but this is related to that question I had this morning in `dev`. Thanks! Oh, are you reviewing in place of @AlasdairSwan?

talbs commented:
@clrux, it looks you addressed all of my HTML/Sass feedback on https://github.com/edx/edx-platform/pull/9367. Thanks for that!

- - -

My understanding was that this was an a11y-minded bug and PR solution, but after chatting it looks like that on top of you resolving the a11y-related DOM and interation issues,  you've made some unrelated visual changes to the navigation

**In the future, please separate out any bug-fix related code changes from extracurricular design revamps so that we can get the bugs squashed and the UI its due review in separate steps. It makes my job as a FED and UX reviewer much easier**

- - -

**Focusing on those UI and Visual Design changes, here are some thoughts: (with screengrabs below all feedback)**

1. Please find out the reason why the proposed link color is different than the current blue. I'd like to not change this and want to know why this is happening (if it was something you added or an inheritance thing).
2. Your branch may be old or you may have manually changed this, but the homework icon symbol has changed recently. Your UI is using the older alarm clock still - please change.
3. Please sync up the ``.menu-item.active`` styling so that it shares the same ``font-weight`` as the  ``.group-heading.active`` (700). Active/current states should share some similar visual characteristics to help with users comprehending what that state means
4. The hit targets for your ``. menu-item`` elements has been reduced. In the current nav they are full-width/block-level elements and easier to select. Please correct.
5. the ``.subtitle`` portion of a ``.menu-item`` should be ~12px, not the 14px your proposed UI has. Please correct
6. I'm okay with removing the slide transition, but it would be good to focus the user's attention on what part of the accordion nav is changing by another transition (using CSS3). Can you please try A) transitioning the ``.chapter.is-open`` state's background color or B) the opacity of a ``.chapter.is-open``'s ``.ui-accordion-content-active`` element from 0.0 to 1.0 (in addition to other semantic ways you're hiding that UI when its collapsed)?
7. The hover/focus state of ``.menu-item`` elements is pretty subtle compared to what is there now (color change vs. border outline). Please try to make this color change/transition be more attention grabbing (by A) starting with a less dark, but still a11y-friendly default color or B) adding a border-bottom/underline).
8. The active/current state of the ``.menu-item.active`` items seems like it would fail color contrast tests (with the darker gray background on dark text). Can you confirm that this passes? If not, please address.

Let me know when these are addressed and I'll continue to review.

- - -

**Proposed LMS Nav**
![lms-proposed](https://cloud.githubusercontent.com/assets/163763/9658148/2a98986a-5215-11e5-8e12-62e34036bde3.png)


**Current LMS Nav**
![lms-current](https://cloud.githubusercontent.com/assets/163763/9658153/2eb81092-5215-11e5-99f4-1eb650a36e34.png)

cptvitamin commented on lms/templates/courseware/courseware.html:
```diff
@@ -186,8 +183,8 @@
-      <div id="accordion" style="display: none">
-        <nav aria-label="${_('Course Navigation')}">
+      <div class="accordion">
+        <nav class="course-navigation" aria-label="${_('Course Navigation')}">
```
@clrux sorry I did not catch this the first time, but there is no need for the word "Navigation" in the aria-label since the screen reader already reports the role of the element, which is "navigation" so I am hearing "Course Navigation navigation"

cptvitamin commented on lms/static/js/fixtures/accordion.html:
```diff
@@ -0,0 +1,56 @@
+<div class="course-wrapper">
+  <div class="accordion">
+    <button class="button-chapter chapter" aria-controls="accordion-menu-1" aria-pressed="true">
+      <h3 class="group-heading">
```
You should probably update this fixture

cptvitamin commented on lms/templates/courseware/accordion.html:
```diff
@@ -6,73 +6,42 @@
+        ${chapter['display_name']}
+    </span>
+</button>
+<div class="chapter-content-container" id="${chapter['display_id']}-child" tabindex="-1" aria-expanded="false" role="region" aria-labelledby="${chapter['display_id']}-parent" aria-hidden="false">
```
@clrux I think I want to use role="group" here. I think its more semantic than region, which implies a much larger area of items that may or may not be related.  This is clearly a group here, a group of links.  On that note, I think we need to make the label more precise. Currently, it doesn't inform the user that its a submenu.  Instead of an aria-labelledby here, can we make it an aria-label that reports the parent name and appends "submenu" so that it gets announced as "Introduction submenu, group"

clrux commented:
jenkins run bokchoy
jenkins run lettuce

clrux commented:
jenkins run bokchoy

clrux commented:
jenkins run bokchoy

talbs commented:
@clrux, what's the status of things? Should I re-review or are you still addressing the [feedback I gave above](https://github.com/edx/edx-platform/pull/9525#issuecomment-137429980)?

clrux commented:
@talbs Just waiting on thumbs from everyone at this point. I believe I still need @AlasdairSwan to review the JS. Your feedback is addressed. I'll get an updated sandy spun up here in a bit. Same URL.

talbs commented:
Okay, I shall await a fresh litterbox, errr, I mean sandbox. Thanks for the update.

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+            listenForClick: function() {
+                $('.accordion').on('click', '.button-chapter', function(event) {
+                    navigation.closeAccordions(event.currentTarget);
+                    navigation.openAccordionSection(event.currentTarget);
```
In both of these functions you are creating a jQuery object from the element so it would make sense to create it only once in this function and pass it to the close and open functions.

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+
+            closeAccordions: function(button) {
+                var menu = $(button).next('.chapter-content-container').find('.chapter-menu');
+                console.log(button, menu);
```
Please remove this debugging code before merging.

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+                console.log(button, menu);
+
+                $('.accordion .button-chapter').each(function(event) {
+                    var el = $(this);
```
For clarity can you please pass the element as the second parameter in the each function and use that instead of 'this'.

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+
+                    el.removeClass('is-open').attr('aria-expanded', 'false');
+                    el.children('.group-heading').removeClass('active');
+                    el.children('.group-heading').find('.icon').addClass('fa-caret-right').removeClass('fa-caret-down');
```
You shouldn't need to get  el.children('.group-heading') twice because of chaining

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+                    el.children('.group-heading').find('.icon').addClass('fa-caret-right').removeClass('fa-caret-down');
+
+                    el.next('.chapter-content-container').removeClass('is-open');
+                    el.next('.chapter-content-container').find('.chapter-menu').not(menu).slideUp();
```
You shouldn't need to get el.next('.chapter-content-container') twice because of chaining

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+
+                $(section).addClass('is-open').attr('aria-expanded', 'true');
+                $(section).children('.group-heading').addClass('active');
+                $(section).children('.group-heading').find('.icon').removeClass('fa-caret-right').addClass('fa-caret-down');
```
You should not be creating the same jQuery object 4 times. As mentioned in a previous comment it would be better to pass the jQuery object into the function so you don't need to create it here at all.

You shouldn't need to get el.children('.group-heading') twice because of chaining. Can you also please have a new line for each method to make it easier to read.

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+            },
+
+            checkForCurrent: function() {
+                var active = $('.accordion .chapter-content-container:has(.active)').index('.accordion .chapter-content-container') ? $('.accordion .chapter-content-container:has(.active)').index('.accordion .chapter-content-container') : 0,
```
This line of code is too long. I think it would be better to have a function called getActiveIndex that spreads out the code over several lines and makes it easier to read.

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,69 @@
+                    activeSection = $('.accordion .button-chapter:eq(' + active + ')');
+
+                navigation.closeAccordions();
+                navigation.openAccordionSection(activeSection);
```
This should only be called if there is an activeSection, so you probably don't need the activeSection variable and can define the element in the openAccordionSection call if active index > 0.

AlasdairSwan commented on lms/static/js/spec/navigation_spec.js:
```diff
@@ -0,0 +1,66 @@
+                it('ensures new section is opened and previous section is closed', function() {
+                    button:eq(1).click();
+
+                    console.log(chapterContent:eq(0), chapterContent:eq(1));
```
Please remove before merging

AlasdairSwan commented on lms/static/js/spec/navigation_spec.js:
```diff
@@ -0,0 +1,66 @@
+                    expect(chapterContent:eq(0)).not.toHaveClass('is-open');
+                    expect(chapterContent:eq(1)).toHaveClass('is-open');
+
+                    console.log(button:eq(0), button:eq(1));
```
Please remove before merging

AlasdairSwan commented on lms/static/js/spec/navigation_spec.js:
```diff
@@ -0,0 +1,66 @@
+            });
+        });
+    });
+});
```
There should probably be tests for closing too.

clrux replied:
Because this is an accordion there is never all closed sections. I have a test that checks to see if, when one section opens, the other closes.

AlasdairSwan commented on lms/templates/instructor/instructor_dashboard_2/e-commerce.html:
```diff
@@ -204,11 +204,11 @@ <h3 class="error-msgs" id="error-msg"></h3>
+                    var active = jQuery(".accordion").accordion('option', 'active');
                     $.cookie('saved_index', null);
                     $.cookie('saved_index', active);
                     $('#error-msg').val('');
```
Out of scope of this PR but I wish I hadn't seen this file :(

clrux commented:
@talbs @AlasdairSwan @cptvitamin I've gone through and addressed feedback from each of you. I am unfortunately unable to spin up a sandbox at this time because there's an issue with the pipeline looking for removed CoffeeScript files. (Andy Armstrong has an in-progress PR to address this on Master). To complete your reviews, please spin this up locally.

clrux commented:
jenkins run lettuce
jenkins run bokchoy

clrux commented:
jenkins run js

talbs commented:
@clrux, things are looking good visually. I did notice that you reverted your change to the ``font-weight`` of``.menu-item`` and ``.group-heading``. I preferred it with what you revised it to (600 rather than 700). Mind changing that back?

talbs commented:
@clrux, I believe you've addressed all of my FED-based feedback. Once you sync up the ``font-weight`` revert comment, I think things are good from my perspective. Once you've squashed your commits into logical ones and everyone else signs off on their bits, this has my :+1:

clrux commented:
jenkins run bokchoy

clrux commented:
@AlasdairSwan and @cptvitamin would you gentleman thumbs if this is to your liking?

AlasdairSwan commented on lms/static/js/utils/navigation.js:
```diff
@@ -0,0 +1,114 @@
+            closeAccordions: function(button, section) {
+                var menu = $(section).find('.chapter-menu'), toggle;
+
+                $('.accordion .button-chapter').each(function(index, event) {
```
event is actually element, please rename so it is easier to read

AlasdairSwan commented:
Once the closeAccordians each function variable has been updated it's good to go :+1:

cptvitamin commented:
@clrux there is no focus indicator on the top level buttons in FF.

clrux commented:
@clrux There's an "="" in the `span` within the `button` chapter.

clrux commented:
@talbs I've got a [fresh sandbox](http://clrux-ac-76.sandbox.edx.org) up with the following changes as per our discussion(s). I've unchecked your name from the completed reviewers list. Please re-review and then check off your name once you approve.

* hover and focus styles for the course navigation button items
* no outlines appear on click

clrux commented:
jenkins run js

talbs commented on lms/static/sass/course/courseware/_sidebar.scss:
```diff
@@ -1,234 +1,178 @@
+                background: $white;
+
+                &:focus {
+                    // we don't need the blue glow to be here at this point
```
Not sure we need this comment here. Who is it helping? If its worth explaining why you're resetting outline, I'd re-word this to focus on 1) why here? and 2) why its okay to do so here?

talbs commented on lms/static/sass/course/courseware/_sidebar.scss:
```diff
@@ -1,234 +1,178 @@
+                &:focus {
+                    // we don't need the blue glow to be here at this point
+                    // although focus is sent for correct keyboard/a11y
+                    outline: none;
```
Should this be made into a Sass extend/helper that 1) we can provide a bit more background on what its used for and 2) we can call from other future UI spaces that we explicitly set it from?

clrux replied:
We have this technically, just with a different name (`.sr-is-focusable`). It's a bit different here though because this container contains something whereas the other doesn't. I think it's fine to be a one-off thing here. If we use this method again in the future we can revisit.

talbs commented:
@clrux, comparing this sandbox to stage, I'm seeing some differences in how the DOM/UI elements are displaying on load. Here's a video with courses.stage.edx.org to the left and your sandbox to the right - http://cl.ly/image/43283j0j3N2u. Watch as I force refresh both windows (your DOM is showing and then collapsing). Why is this and can you return to stage's behavior (assuming it is not a detriment to a11y)?

talbs commented:
@clrux, more visual feedback...

![image](https://cloud.githubusercontent.com/assets/163763/10003370/0cde4db0-607a-11e5-9ec1-3e0f66f235c5.png)

The arrows to the left of the sections appear like they are both bigger and darker than what's on stage right now. Is this on purpose and needed for a11y improvements/issues?

clrux commented:
jenkins run python

clrux commented:
@cptvitamin and @talbs please [continue your reviews](http://clrux-ac-76.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/).

talbs commented:
@clrux, a few more visual notes from your updated sandbox...

**Subsection Hover State**
Please make sure the hover states of subsections remains the same color as the default state (they're currently turning blue (perhaps from a default/base link style). Also, please make the new hover background styling that you've added a bit lighter than the current item's background color (perhaps ``$gray-l5`` since the active state is ``$gray-l4``. See screengrab below for ideal hover state with the above in mind:

![image](https://cloud.githubusercontent.com/assets/163763/10027501/5692380c-6135-11e5-950a-a417a88bd6ac.png)

- - -

**Focus still showing on expanded sections**
When I click on a section to expand it, the native focus styling is still being applied (see screengrab below). This should not happen, right?

![image](https://cloud.githubusercontent.com/assets/163763/10027522/7f191462-6135-11e5-9fba-b42012cec54c.png)

clrux commented:
@talbs I've addressed the hover color on the individual sections. The `:active` styles were working, but because we changed how focus was being sent, it stayed on the button as `:focus`, which we need. However, I've changed the `button`'s to `a role="button"` which serves the same purpose but handles focus differently. You'll now be able to `Tab` through and see focus, but if you click, you won't. This should satisfy the remainder of your feedback.

@cptvitamin let me know if you're okay with the above change and if you have any other feedback.

The sandbox has been updated as of 22 Sep, 15:51.

talbs commented:
@clrux, thanks for the updates. The visuals look good (assuming they are consistent across browsers). Please check that, squash your commits, and make sure tests pass.

:+1:

clrux commented:
@cptvitamin I've added JavaScript to handle Spacebar keypresses on the "buttons" so they function as expected as if they were actually buttons. The sandbox is updated. Please review and let me know your thoughts.

clrux commented:
jenkins run lettuce

clrux commented:
I have verbal thumbs from @cptvitamin.

andy-armstrong commented on lms/static/js/spec/navigation_spec.js:
```diff
@@ -0,0 +1,93 @@
+                it('ensures new section is opened and previous section is closed', function() {
+                    button:eq(1).click();
+
+                    expect(chapterContent:eq(0)).not.toHaveClass('is-open');
```
@clrux @ahsan-ul-haq This code is not valid JavaScript, so the test is silently failing. It looks like this was CoffeeScript that didn't get converted correctly.

@benpatterson Any idea why the JSHint wouldn't have found this? It would also be good to get js-test-tools to fail for invalid test files, rather than silently ignoring them.

benpatterson replied:
@andy-armstrong Actually if my memory serves me correctly, this PR is what boosted our JSHint errors into the 'warm' zone on our dashboard. Indeed, these JsHint violations were missed due to a bug in how we are running diff-quality. I fixed that in https://github.com/edx/edx-platform/pull/9898.

CC @clrux @ahsan-ul-haq

andy-armstrong replied:
Excellent @benpatterson. I'm glad to hear we have that covered now. Can someone please fix the tests so that they are still running please.
